src/main/java/picard/illumina/ClusterDataToSamConverter.java

```diff
@@ -207,49 +241,86 @@ private SAMRecord createSamRecord(final ReadData readData, final String readName
+        if (unmatchedBarcode != null && includeQualitiesWithBarcode) {
+            barcodeQuality = getBarcodeQuality(cluster);
+        }
+
+
+        final List<String> molecularIndexes = new ArrayList<>();
+        final List<String> molecularIndexQualities = new ArrayList<>();
         if (hasMolecularBarcode) {
-            molecularIndexes        = new ArrayList<>();
-            molecularIndexQualities = new ArrayList<>();
-            for (int i = 0; i < molecularBarcodeIndices.length; i++) {
-                molecularIndexes.add(new String(cluster.getRead(molecularBarcodeIndices[i]).getBases()).replace('.', 'N'));
-                molecularIndexQualities.add(SAMUtils.phredToFastq(cluster.getRead(molecularBarcodeIndices[i]).getQualities()));
+            for (int molecularBarcodeIndice : molecularBarcodeIndices) {
+                molecularIndexes.add(new String(cluster.getRead(molecularBarcodeIndice).getBases()).replace('.', 'N'));
+                molecularIndexQualities.add(SAMUtils.phredToFastq(cluster.getRead(molecularBarcodeIndice).getQualities()));
             }
-        } else {
-            molecularIndexes        = Collections.emptyList();
-            molecularIndexQualities = Collections.emptyList();
         }
 
         final SAMRecord firstOfPair = createSamRecord(
-            cluster.getRead(templateIndices[0]), readName, cluster.isPf(), true, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+                cluster.getRead(templateIndices[0]),
+                readName,
+                cluster.isPf(),
+                true,
+                unmatchedBarcode,
+                barcodeQuality,
+                molecularIndexes,
+                molecularIndexQualities);
         ret.records[0] = firstOfPair;
 
+
         SAMRecord secondOfPair = null;
 
-        if(isPairedEnd) {
-            secondOfPair  = createSamRecord(
-                cluster.getRead(templateIndices[1]), readName, cluster.isPf(), false, unmatchedBarcode, molecularIndexes, molecularIndexQualities);
+        if (isPairedEnd) {
+            secondOfPair = createSamRecord(
+                    cluster.getRead(templateIndices[1]),
+                    readName,
+                    cluster.isPf(),
+                    false,
+                    unmatchedBarcode,
+                    barcodeQuality,
+                    molecularIndexes,
+                    molecularIndexQualities);
             ret.records[1] = secondOfPair;
         }
 
         if (adapterMarker != null) {
             // Clip the read
             if (isPairedEnd) {
                 adapterMarker.adapterTrimIlluminaPairedReads(firstOfPair, secondOfPair);
-            }
-            else {
+            } else {
                 adapterMarker.adapterTrimIlluminaSingleRead(firstOfPair);
             }
         }
         return ret;
     }
+
+    private String getBarcodeQuality(ClusterData cluster) {
+        final StringJoiner barcodeQ = new StringJoiner("~");
```

Comment:
Can this magic string be replaced with a declared constant? I see `SamToFastqWithTags.TAG_SPLIT_QUAL`.

Reply:
it's private...I'll put something in htsjdk.